src/components/transport_manager/src/transport_adapter/transport_adapter_impl.cc

```diff
@@ -836,15 +858,26 @@ DeviceSptr TransportAdapterImpl::FindDevice(const DeviceUID& device_id) const {
 
 void TransportAdapterImpl::ConnectPending(const DeviceUID& device_id,
                                           const ApplicationHandle& app_handle) {
-  connections_lock_.AcquireForReading();
+  connections_lock_.AcquireForWriting();
   ConnectionMap::iterator it_conn =
       connections_.find(std::make_pair(device_id, app_handle));
   if (it_conn != connections_.end()) {
     ConnectionInfo& info = it_conn->second;
+    if (info.state == ConnectionInfo::ESTABLISHED) {
+      LOG4CXX_DEBUG(logger_, "Reset Connection Pointer");
+      connections_.erase(std::make_pair(device_id, app_handle));
+
+      connections_lock_.Release();
+      CreateDevice(device_id);
+      connections_lock_.AcquireForWriting();
```

Comment:
This is pretty ugly, perhaps just set a flag and `CreateDevice` after you're done writing.